_posts/2019-05-10-server-rendering-responsively.md

```diff
@@ -0,0 +1,189 @@
+
+There's one type of challenge with building a responsive app that `styled-system` can't solve: when we need to emit
+different layouts across different breakpoints. In this case, we need something that can render very different
+component sub-trees. We couldn't find an approach that satisfied our needs, so we wrote our own.
+
+## Tool 2: [`@artsy/react-responsive-media`](https://github.com/artsy/react-responsive-media)
+
+First off, an announcement: we've just released `react-responsive-media` version 1.0! (TODO: release v1.0 😬 or
+change this announcement)
+
+`react-responsive-media` allows you to define a set of breakpoint widths, then declaratively render component
+sub-trees when those breakpoints are met. It looks something like this:
+
+```xml
+<>
+  <Media at="xs">
+    <MobileLayout />
+  </Media>
+  <Media greaterThan="xs">
+    <NonMobileLayout />
+  </Media>
+</>
+```
+
+In this example, we're emitting the `MobileLayout` component for devices at or below our `xs` breakpoint, and the
+`NonMobileLayout` for devices greater than our `xs` breakpoint. You can imagine that the `MobileLayout` and
+`NonMobileLayout` components are complicated sub-trees, with more significant differences than `styled-system`
+could handle.
+
+### How it works
+
+The first important thing to note is that when server-rendering with `react-responsive-media`, **all** breakpoints
+get rendered by the server.
+
+Why not just the one that the current device needs? A couple reasons. First, we can't _accurately_ identify which
+breakpoint your device needs from the server. We could use a library to sniff the browser `user-agent`, but those
+aren't always accurate, and they wouldn't give us all the information we need to know when we are server-rendering.
+
+The other reason `react-responsive-media` renders _all_ breakpoints has to do with the hydration of a
+server-rendered React app. When using SSR in a React app, you need to call
+[`ReactDOM.hydrate()`](https://reactjs.org/docs/react-dom.html#hydrate) to connect your server-rendered app to the
+client's browser. As the docs describe,
+
+> React expects that the rendered content is identical between the server and the client.
+
+If there are differences between your server-rendered component tree and the initial client-rendered component
+tree, a couple things will happen.
+
+1. You'll get a friendly error in your console telling you there were differences (in development mode), but
+   also...
+2. weird, unpredictable things will happen to your app.
+
+Here's what our app looked like in one case where our client render didn't match our server render:
+
+(TODO: image of spinning body)
+
+Yuck! The docs explain that `ReactDOM.hydrate` doesn't put a lot of effort into patching up differences between the
+server and client renders, and that's how you end up with weird things like above.
+
+Because hydration is so picky about the component trees matching between server and client, we render _all_
```

Comment:
It's not picky

Reply:
Just based off this comment I'm not sure how this sentence could be improved. @damassi do you have a specific recommendation about wording or a sense of what idea you'd like this to convey?

Reply:
As far as I understand it, React's hydration mechanism is unrelated to us rendering out all breakpoints; we render everything so that we can use native CSS media matching to instantly deal with intermediate and / or ambiguous sizes, such as "medium" (not mobile, not desktop, but also not a tablet, etc) -- in other words, so that we're not constrained by screen-size to user-agent and can cover each and every design possibility without the need for JS.

Reply:
I'm happy to use a different word than "picky".

My interpretation from talking to people is that it _is_ significant that React's hydration algorithm is naive & assumes your server/client matches; and it _did_ contribute to us choosing to render all breakpoints, so that we didn't have to deal with mismatched subtrees. 

I defer to you all, but that's what I've learned in the research for this post.

Reply:
Yah, we're outputting all breakpoints are so that we can render visually correct layouts at all screen dimensions, regardless of user-agent, from the server render pass down to the client-side handshake / hydration. It's the secret ingredient, and where the innovation around this idea is centered, compared to other approaches. @alloy can comment more, however.

Reply:
Here's my [initial feedback](https://github.com/artsy/reaction/issues/1367) on why we had to move from responsive to a `react-responsive-media` like approach. 

@pepopowitz I don't think I noticed this conversation included in the post, but I'd recommend linking off to it. https://github.com/artsy/reaction/issues/1367

Reply:
I reworded to remove "picky", and added a link to the inspiring PR.

Reply:
1. I reworded without "picky" - https://github.com/artsy/artsy.github.io/pull/566/files#diff-a879e573d311218fd7eec7da14d776f5R118
2. I linked to the inspiring PR - https://github.com/artsy/artsy.github.io/pull/566/files#diff-a879e573d311218fd7eec7da14d776f5R122